src/network/NetworkKey.cpp

```diff
@@ -0,0 +1,444 @@
+#pragma region Copyright (c) 2014-2016 OpenRCT2 Developers
+/*****************************************************************************
+ * OpenRCT2, an open source clone of Roller Coaster Tycoon 2.
+ *
+ * OpenRCT2 is the work of many authors, a full list can be found in contributors.md
+ * For more information, visit https://github.com/OpenRCT2/OpenRCT2
+ *
+ * OpenRCT2 is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * A full copy of the GNU General Public License can be found in licence.txt
+ *****************************************************************************/
+#pragma endregion
+
+#ifndef DISABLE_NETWORK
+
+#include "NetworkKey.h"
+#include "../diagnostic.h"
+
+#include <openssl/evp.h>
+#include <openssl/rsa.h>
+#include <openssl/pem.h>
+#include <vector>
+
+#define KEY_TYPE EVP_PKEY_RSA
+
+constexpr int KEY_LENGTH_BITS = 2048;
+
+NetworkKey::NetworkKey()
+{
+    _ctx = EVP_PKEY_CTX_new_id(KEY_TYPE, nullptr);
+    if (_ctx == nullptr)
+    {
+        log_error("Failed to create OpenSSL context");
+    }
+}
+
+NetworkKey::~NetworkKey()
+{
+    Unload();
+    if (_ctx != nullptr)
+    {
+        EVP_PKEY_CTX_free(_ctx);
+        _ctx = nullptr;
+    }
+}
+
+void NetworkKey::Unload()
+{
+    if (_key != nullptr)
+    {
+        EVP_PKEY_free(_key);
+        _key = nullptr;
+    }
+}
+
+bool NetworkKey::Generate()
+{
+    if (_ctx == nullptr)
+    {
+        log_error("Invalid OpenSSL context");
+        return false;
+    }
+#if KEY_TYPE == EVP_PKEY_RSA
+    if (!EVP_PKEY_CTX_set_rsa_keygen_bits(_ctx, KEY_LENGTH_BITS))
+    {
+        log_error("Failed to set keygen params");
+        return false;
+    }
+#else
+    #error Only RSA is supported!
+#endif
+    if (EVP_PKEY_keygen_init(_ctx) <= 0)
+    {
+        log_error("Failed to initialise keygen algorithm");
+        return false;
+    }
+    if (EVP_PKEY_keygen(_ctx, &_key) <= 0)
+    {
+        log_error("Failed to generate new key!");
+        return false;
+    }
+    else
+    {
+        log_verbose("Key successfully generated");
+    }
+    log_verbose("New key of type %d, length %d generated successfully.", KEY_TYPE, KEY_LENGTH_BITS);
+    return true;
+}
+
+bool NetworkKey::LoadPrivate(SDL_RWops * file)
+{
+    size_t size = (size_t)file->size(file);
+    if (size == (size_t)-1)
+    {
+        log_error("unknown size, refusing to load key");
+        return false;
+    }
+    else if (size > 4 * 1024 * 1024)
+    {
+        log_error("Key file suspiciously large, refusing to load it");
+        return false;
+    }
+    char * priv_key = new char[size];
```

Comment:
Yes, when the size of data is unknown beforehand, I use vector, so I can easily resize it later on.